File: src-tauri/src/domain/workspace/value_objects/workspace_context.rs
use crate::domain::project::value_objects::ProjectId;
use crate::domain::workspace::errors::WorkspaceError;
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};

/// WorkspaceContext represents the context and state of an active project workspace
///
/// This value object contains immutable information about a workspace session,
/// including the project being worked on and the current navigation state.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WorkspaceContext {
    /// The unique identifier of the project this workspace represents
    project_id: ProjectId,
    /// The human-readable name of the project
    project_name: String,
    /// The root source folder path for this project
    source_folder: PathBuf,
    /// The current navigation path within the workspace
    current_path: PathBuf,
}

impl WorkspaceContext {
    /// Create a new WorkspaceContext
    ///
    /// # Arguments
    /// * `project_id` - The unique identifier of the project
    /// * `project_name` - The human-readable name of the project
    /// * `source_folder` - The root source folder path
    /// * `current_path` - The current navigation path (defaults to source_folder if None)
    ///
    /// # Errors
    /// Returns `WorkspaceError` if:
    /// - Project name is empty
    /// - Source folder path is invalid
    /// - Current path is outside the source folder boundary
    pub fn new(
        project_id: ProjectId,
        project_name: impl Into<String>,
        source_folder: impl AsRef<Path>,
        current_path: Option<impl AsRef<Path>>,
    ) -> Result<Self, WorkspaceError> {
        let project_name = project_name.into();
        let source_folder = source_folder.as_ref().to_path_buf();
        let current_path = current_path
            .map(|p| p.as_ref().to_path_buf())
            .unwrap_or_else(|| source_folder.clone());

        // Validate project name
        if project_name.trim().is_empty() {
            return Err(WorkspaceError::invalid_workspace_context(
                "Project name cannot be empty",
            ));
        }

        // Validate paths
        if !source_folder.is_absolute() {
            return Err(WorkspaceError::invalid_workspace_context(
                "Source folder must be an absolute path",
            ));
        }

        // Ensure current path is within source folder boundaries
        if !Self::is_path_within_boundary(&current_path, &source_folder)? {
            return Err(WorkspaceError::navigation_boundary_violation(
                current_path.display().to_string(),
                source_folder.display().to_string(),
            ));
        }

        Ok(WorkspaceContext {
            project_id,
            project_name,
            source_folder,
            current_path,
        })
    }

    /// Get the project ID
    pub fn project_id(&self) -> &ProjectId {
        &self.project_id
    }

    /// Get the project name
    pub fn project_name(&self) -> &str {
        &self.project_name
    }

    /// Get the source folder path
    pub fn source_folder(&self) -> &Path {
        &self.source_folder
    }

    /// Get the current path
    pub fn current_path(&self) -> &Path {
        &self.current_path
    }

    /// Check if the current path is at the workspace root
    pub fn is_at_root(&self) -> bool {
        self.current_path == self.source_folder
    }

    /// Get the parent path if navigation up is possible
    ///
    /// Returns `None` if already at the workspace root
    pub fn get_parent_path(&self) -> Option<PathBuf> {
        if self.is_at_root() {
            None
        } else {
            self.current_path.parent().map(|p| p.to_path_buf())
        }
    }

    /// Create a new WorkspaceContext with updated current path
    ///
    /// # Arguments
    /// * `new_path` - The new current path
    ///
    /// # Errors
    /// Returns `WorkspaceError` if the new path is outside workspace boundaries
    pub fn with_current_path(&self, new_path: impl AsRef<Path>) -> Result<Self, WorkspaceError> {
        let new_path = new_path.as_ref().to_path_buf();

        if !Self::is_path_within_boundary(&new_path, &self.source_folder)? {
            return Err(WorkspaceError::navigation_boundary_violation(
                new_path.display().to_string(),
                self.source_folder.display().to_string(),
            ));
        }

        Ok(WorkspaceContext {
            project_id: self.project_id.clone(),
            project_name: self.project_name.clone(),
            source_folder: self.source_folder.clone(),
            current_path: new_path,
        })
    }

    /// Navigate to a child folder
    ///
    /// # Arguments
    /// * `folder_name` - Name of the folder to navigate to (relative to current path)
    ///
    /// # Errors
    /// Returns `WorkspaceError` if:
    /// - Folder name contains invalid characters
    /// - Navigation would exceed workspace boundaries
    pub fn navigate_to_folder(&self, folder_name: &str) -> Result<Self, WorkspaceError> {
        // Validate folder name for security
        if folder_name.contains("..") || folder_name.contains('/') || folder_name.contains('\\') {
            return Err(WorkspaceError::invalid_path(
                folder_name.to_string(),
                "Folder name contains invalid characters or path traversal",
            ));
        }

        if folder_name.trim().is_empty() {
            return Err(WorkspaceError::invalid_path(
                folder_name.to_string(),
                "Folder name cannot be empty",
            ));
        }

        let new_path = self.current_path.join(folder_name);
        self.with_current_path(new_path)
    }

    /// Navigate to parent directory
    ///
    /// # Errors
    /// Returns `WorkspaceError` if already at workspace root
    pub fn navigate_to_parent(&self) -> Result<Self, WorkspaceError> {
        match self.get_parent_path() {
            Some(parent_path) => self.with_current_path(parent_path),
            None => Err(WorkspaceError::invalid_path(
                self.current_path.display().to_string(),
                "Already at workspace root, cannot navigate up",
            )),
        }
    }

    /// Check if a path is within the workspace boundary
    ///
    /// This is a critical security function that prevents path traversal attacks
    fn is_path_within_boundary(path: &Path, boundary: &Path) -> Result<bool, WorkspaceError> {
        // Canonicalize both paths to resolve any symlinks or relative components
        let canonical_path = path
            .canonicalize()
            .or_else(|_| {
                // If path doesn't exist yet, canonicalize the parent and join the filename
                if let Some(parent) = path.parent() {
                    if let Some(filename) = path.file_name() {
                        parent
                            .canonicalize()
                            .map(|canonical_parent| canonical_parent.join(filename))
                    } else {
                        Err(std::io::Error::new(
                            std::io::ErrorKind::InvalidInput,
                            "Invalid path",
                        ))
                    }
                } else {
                    Err(std::io::Error::new(
                        std::io::ErrorKind::InvalidInput,
                        "Invalid path",
                    ))
                }
            })
            .map_err(|e| {
                WorkspaceError::invalid_path(
                    path.display().to_string(),
                    format!("Failed to canonicalize path: {}", e),
                )
            })?;

        let canonical_boundary = boundary.canonicalize().map_err(|e| {
            WorkspaceError::source_folder_not_found(format!(
                "Workspace boundary path invalid: {}",
                e
            ))
        })?;

        Ok(canonical_path.starts_with(canonical_boundary))
    }

    /// Get relative path from workspace root
    pub fn relative_path(&self) -> Result<PathBuf, WorkspaceError> {
        self.current_path
            .strip_prefix(&self.source_folder)
            .map(|p| p.to_path_buf())
            .map_err(|_| {
                WorkspaceError::invalid_workspace_context(
                    "Current path is not within workspace source folder",
                )
            })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn create_test_project_id() -> ProjectId {
        ProjectId::new()
    }

    #[test]
    fn test_workspace_context_creation() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        let context =
            WorkspaceContext::new(project_id.clone(), "Test Project", temp_dir.path(), None::<&str>)
                .unwrap();

        assert_eq!(context.project_id(), &project_id);
        assert_eq!(context.project_name(), "Test Project");
        assert_eq!(context.source_folder(), temp_dir.path());
        assert_eq!(context.current_path(), temp_dir.path());
        assert!(context.is_at_root());
    }

    #[test]
    fn test_empty_project_name_validation() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        let result = WorkspaceContext::new(project_id, "", temp_dir.path(), None::<&str>);

        assert!(result.is_err());
        match result.unwrap_err() {
            WorkspaceError::InvalidWorkspaceContext { reason } => {
                assert!(reason.contains("Project name cannot be empty"));
            }
            _ => panic!("Expected InvalidWorkspaceContext error"),
        }
    }

    #[test]
    fn test_navigation_to_folder() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        // Create a subdirectory
        let sub_dir = temp_dir.path().join("documents");
        std::fs::create_dir(&sub_dir).unwrap();

        let context = WorkspaceContext::new(
            project_id,
            "Test Project",
            temp_dir.path(),
            None::<&str>,
        )
        .unwrap();

        let navigated_context = context.navigate_to_folder("documents").unwrap();

        assert_eq!(navigated_context.current_path(), sub_dir);
        assert!(!navigated_context.is_at_root());
        assert!(navigated_context.get_parent_path().is_some());
    }

    #[test]
    fn test_navigation_to_parent() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        // Create a subdirectory
        let sub_dir = temp_dir.path().join("documents");
        std::fs::create_dir(&sub_dir).unwrap();

        let context =
            WorkspaceContext::new(project_id, "Test Project", temp_dir.path(), Some(&sub_dir))
                .unwrap();

        let parent_context = context.navigate_to_parent().unwrap();

        assert_eq!(parent_context.current_path(), temp_dir.path());
        assert!(parent_context.is_at_root());
    }

    #[test]
    fn test_navigation_boundary_violation() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        let context = WorkspaceContext::new(
            project_id,
            "Test Project",
            temp_dir.path(),
            None::<&str>,
        )
        .unwrap();

        // Try to navigate with path traversal
        let result = context.navigate_to_folder("../evil");
        assert!(result.is_err());

        match result.unwrap_err() {
            WorkspaceError::InvalidPath { path, reason } => {
                assert_eq!(path, "../evil");
                assert!(reason.contains("invalid characters") || reason.contains("path traversal"));
            }
            _ => panic!("Expected InvalidPath error"),
        }
    }

    #[test]
    fn test_navigation_to_parent_at_root() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        let context = WorkspaceContext::new(
            project_id,
            "Test Project",
            temp_dir.path(),
            None::<&str>,
        )
        .unwrap();

        let result = context.navigate_to_parent();
        assert!(result.is_err());

        match result.unwrap_err() {
            WorkspaceError::InvalidPath { reason, .. } => {
                assert!(reason.contains("Already at workspace root"));
            }
            _ => panic!("Expected InvalidPath error"),
        }
    }

    #[test]
    fn test_relative_path() {
        let temp_dir = TempDir::new().unwrap();
        let project_id = create_test_project_id();

        // Create subdirectories
        let sub_dir = temp_dir.path().join("documents").join("archived");
        std::fs::create_dir_all(&sub_dir).unwrap();

        let context =
            WorkspaceContext::new(project_id, "Test Project", temp_dir.path(), Some(&sub_dir))
                .unwrap();

        let relative = context.relative_path().unwrap();
        assert_eq!(relative, PathBuf::from("documents").join("archived"));
    }
}

File: frontend/src/features/workspace/domain/value-objects/workspace-context.ts
import { ProjectId } from "@/features/project-management/types/value-objects";

/**
 * WorkspaceContext represents the context and state of an active project workspace
 *
 * This value object contains immutable information about a workspace session,
 * including the project being worked on and the current navigation state.
 *
 * Business Rules:
 * - Project ID must be valid
 * - Project name cannot be empty
 * - Source folder must be a valid path
 * - Current path must be within source folder boundary
 * - Context is immutable - changes create new instances
 */
export class WorkspaceContext {
  private constructor(
    private readonly _projectId: ProjectId,
    private readonly _projectName: string,
    private readonly _sourceFolder: string,
    private readonly _currentPath: string
  ) {
    this.validateProjectName(_projectName);
    this.validateSourceFolder(_sourceFolder);
    this.validateCurrentPath(_currentPath, _sourceFolder);
  }

  /**
   * Create a new WorkspaceContext
   *
   * @param projectId - The unique identifier of the project
   * @param projectName - The human-readable name of the project
   * @param sourceFolder - The root source folder path
   * @param currentPath - The current navigation path (defaults to source folder if not provided)
   * @returns New WorkspaceContext instance
   * @throws WorkspaceContextError if validation fails
   */
  static create(
    projectId: ProjectId,
    projectName: string,
    sourceFolder: string,
    currentPath?: string
  ): WorkspaceContext {
    const contextCurrentPath = currentPath || sourceFolder;
    return new WorkspaceContext(
      projectId,
      projectName,
      sourceFolder,
      contextCurrentPath
    );
  }

  /**
   * Create WorkspaceContext from plain object (for serialization/deserialization)
   */
  static fromPlainObject(data: {
    projectId: string;
    projectName: string;
    sourceFolder: string;
    currentPath: string;
  }): WorkspaceContext {
    const projectId = ProjectId.fromString(data.projectId);
    return new WorkspaceContext(
      projectId,
      data.projectName,
      data.sourceFolder,
      data.currentPath
    );
  }

  /**
   * Get the project ID
   */
  get projectId(): ProjectId {
    return this._projectId;
  }

  /**
   * Get the project name
   */
  get projectName(): string {
    return this._projectName;
  }

  /**
   * Get the source folder path
   */
  get sourceFolder(): string {
    return this._sourceFolder;
  }

  /**
   * Get the current path
   */
  get currentPath(): string {
    return this._currentPath;
  }

  /**
   * Check if the current path is at the workspace root
   */
  get isAtRoot(): boolean {
    return this.normalizedCurrentPath === this.normalizedSourceFolder;
  }

  /**
   * Get the parent path if navigation up is possible
   * Returns null if already at the workspace root
   */
  get parentPath(): string | null {
    if (this.isAtRoot) {
      return null;
    }

    // Get parent directory
    const pathParts = this._currentPath.replace(/\/+$/, "").split("/");
    if (pathParts.length <= 1) {
      return null;
    }

    const parentPath = pathParts.slice(0, -1).join("/") || "/";

    // Ensure parent is still within workspace
    if (!this.isPathWithinWorkspace(parentPath)) {
      return null;
    }

    return parentPath;
  }

  /**
   * Create a new WorkspaceContext with updated current path
   *
   * @param newPath - The new current path
   * @returns New WorkspaceContext instance with updated path
   * @throws WorkspaceContextError if the new path is outside workspace boundaries
   */
  withCurrentPath(newPath: string): WorkspaceContext {
    return new WorkspaceContext(
      this._projectId,
      this._projectName,
      this._sourceFolder,
      newPath
    );
  }

  /**
   * Navigate to a child folder
   *
   * @param folderName - Name of the folder to navigate to (relative to current path)
   * @returns New WorkspaceContext instance with updated path
   * @throws WorkspaceContextError if folder name is invalid or navigation would exceed boundaries
   */
  navigateToFolder(folderName: string): WorkspaceContext {
    this.validateFolderName(folderName);

    const newPath = this.joinPaths(this._currentPath, folderName);
    return this.withCurrentPath(newPath);
  }

  /**
   * Navigate to parent directory
   *
   * @returns New WorkspaceContext instance with parent path
   * @throws WorkspaceContextError if already at workspace root
   */
  navigateToParent(): WorkspaceContext {
    const parent = this.parentPath;
    if (parent === null) {
      throw new WorkspaceContextError(
        "Already at workspace root, cannot navigate up",
        "current_path"
      );
    }

    return this.withCurrentPath(parent);
  }

  /**
   * Get relative path from workspace root
   */
  get relativePath(): string {
    if (this.isAtRoot) {
      return "";
    }

    const normalizedSource = this.normalizedSourceFolder;
    const normalizedCurrent = this.normalizedCurrentPath;

    if (normalizedCurrent.startsWith(normalizedSource)) {
      const relative = normalizedCurrent.substring(normalizedSource.length);
      return relative.startsWith("/") ? relative.substring(1) : relative;
    }

    return "";
  }

  /**
   * Convert to plain object for serialization
   */
  toPlainObject(): {
    projectId: string;
    projectName: string;
    sourceFolder: string;
    currentPath: string;
  } {
    return {
      projectId: this._projectId.value,
      projectName: this._projectName,
      sourceFolder: this._sourceFolder,
      currentPath: this._currentPath,
    };
  }

  /**
   * Check equality with another WorkspaceContext
   */
  equals(other: WorkspaceContext): boolean {
    return (
      this._projectId.equals(other._projectId) &&
      this._projectName === other._projectName &&
      this._sourceFolder === other._sourceFolder &&
      this._currentPath === other._currentPath
    );
  }

  // Private validation and utility methods

  private validateProjectName(projectName: string): void {
    if (!projectName || projectName.trim().length === 0) {
      throw new WorkspaceContextError(
        "Project name cannot be empty",
        "project_name"
      );
    }
  }

  private validateSourceFolder(sourceFolder: string): void {
    if (!sourceFolder || sourceFolder.trim().length === 0) {
      throw new WorkspaceContextError(
        "Source folder cannot be empty",
        "source_folder"
      );
    }

    // Basic path validation - should be absolute
    if (!sourceFolder.startsWith("/")) {
      throw new WorkspaceContextError(
        "Source folder must be an absolute path",
        "source_folder"
      );
    }
  }

  private validateCurrentPath(currentPath: string, sourceFolder: string): void {
    if (!currentPath || currentPath.trim().length === 0) {
      throw new WorkspaceContextError(
        "Current path cannot be empty",
        "current_path"
      );
    }

    if (!this.isPathWithinWorkspace(currentPath)) {
      throw new WorkspaceContextError(
        `Navigation boundary violation: path '${currentPath}' is outside workspace root '${sourceFolder}'`,
        "current_path"
      );
    }
  }

  private validateFolderName(folderName: string): void {
    if (!folderName || folderName.trim().length === 0) {
      throw new WorkspaceContextError(
        "Folder name cannot be empty",
        "folder_name"
      );
    }

    // Check for path traversal attempts
    if (
      folderName.includes("..") ||
      folderName.includes("/") ||
      folderName.includes("\\")
    ) {
      throw new WorkspaceContextError(
        "Folder name contains invalid characters or path traversal",
        "folder_name"
      );
    }
  }

  private isPathWithinWorkspace(path: string): boolean {
    const normalizedPath = this.normalizePath(path);
    const normalizedSource = this.normalizedSourceFolder;

    return (
      normalizedPath === normalizedSource ||
      normalizedPath.startsWith(normalizedSource + "/")
    );
  }

  private normalizePath(path: string): string {
    return path.replace(/\/+/g, "/").replace(/\/$/, "") || "/";
  }

  private get normalizedSourceFolder(): string {
    return this.normalizePath(this._sourceFolder);
  }

  private get normalizedCurrentPath(): string {
    return this.normalizePath(this._currentPath);
  }

  private joinPaths(basePath: string, childPath: string): string {
    const normalizedBase = this.normalizePath(basePath);
    const normalizedChild = childPath.trim();

    if (normalizedBase === "/") {
      return `/${normalizedChild}`;
    }

    return `${normalizedBase}/${normalizedChild}`;
  }
}

/**
 * Error class for WorkspaceContext validation failures
 */
export class WorkspaceContextError extends Error {
  constructor(message: string, public readonly field: string) {
    super(message);
    this.name = "WorkspaceContextError";
  }
}

// Type alias for external API consistency
export interface WorkspaceContextData {
  projectId: string;
  projectName: string;
  sourceFolder: string;
  currentPath: string;
}

// Utility functions for working with WorkspaceContext

/**
 * Type guard to check if an object is a valid WorkspaceContextData
 */
export function isWorkspaceContextData(obj: any): obj is WorkspaceContextData {
  return (
    obj &&
    typeof obj === "object" &&
    typeof obj.projectId === "string" &&
    typeof obj.projectName === "string" &&
    typeof obj.sourceFolder === "string" &&
    typeof obj.currentPath === "string"
  );
}

/**
 * Create WorkspaceContext from unknown data with validation
 */
export function createWorkspaceContextFromData(
  data: unknown
): WorkspaceContext {
  if (!isWorkspaceContextData(data)) {
    throw new WorkspaceContextError(
      "Invalid workspace context data format",
      "data"
    );
  }

  return WorkspaceContext.fromPlainObject(data);
}
